storage/src/lib.rs

```diff
@@ -0,0 +1,5012 @@
+    /// metageneration does not match the given value.
+    pub fn if_metageneration_not_match(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("if_metageneration_not_match".into(), vec![value.into()]);
+        self
+    }
+    /// Resource name of the Cloud KMS key, of the form
+    /// projects/my-project/locations/global/keyRings/my-kr/cryptoKeys/my-key,
+    ///  that will be used to encrypt the object. Overrides the object
+    /// metadata's kms_key_name value, if any.
+    pub fn kms_key_name(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("kms_key_name".into(), vec![value.into()]);
+        self
+    }
+    /// Name of the object. Required when the object metadata is not
+    /// otherwise provided. Overrides the object metadata's name value, if
+    /// any. For information about how to URL encode object names to be path
+    /// safe, see Encoding URI Path Parts.
+    pub fn name(mut self, value: impl Into<String>) -> Self {
+        self.url_params.insert("name".into(), vec![value.into()]);
+        self
+    }
+    /// Apply a predefined set of access controls to this object.
+    pub fn predefined_acl(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("predefined_acl".into(), vec![value.into()]);
+        self
+    }
+    /// Set of properties to return. Defaults to noAcl, unless the object
+    /// resource specifies the acl property, when it defaults to full.
+    pub fn projection(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("projection".into(), vec![value.into()]);
+        self
+    }
+    /// The project to be billed for this request if the target bucket is
+    /// requester-pays bucket.
+    pub fn provisional_user_project(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("provisional_user_project".into(), vec![value.into()]);
+        self
+    }
+    /// The project to be billed for this request. Required for Requester
+    /// Pays buckets.
+    pub fn user_project(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("user_project".into(), vec![value.into()]);
+        self
+    }
+    /// Explicitly sets the content type of the media being uploaded.
+    pub fn media_content_type(mut self, value: impl Into<String>) -> Self {
+        self.media_content_type = Some(value.into());
+        self
+    }
+
+    pub async fn upload<'a, R>(mut self, reader: &'a mut R) -> Result<model::Object>
+    where
+        R: tokio::io::AsyncRead + Unpin,
+    {
+        let client = self.client.inner;
+        let tok = client.cred.access_token().await.map_err(Error::wrap)?;
+        let body = serde_json::to_vec(&self.request).map_err(Error::wrap)?;
+        let mut buf: Vec<u8> = vec![];
+        tokio::io::copy(reader, &mut buf)
```

Comment:
It looks like the the entire payload is loaded into memory.  Is my understanding correct?

Reply:
For this uploading and downloading, I'm surprised that reqwest doesn't accept the standard `AsyncRead` and `AsyncWrite` types.

I don't think we can solve this problem really well without feedback from the larger community, so I propose a **new strategy**:  Expose all the reqwest types in our APIs.  It's the easiest thing to do, no?  Let our early adopters tell us if it works for them or not.

Reply:
I think what we now have for downloads and uploads should be good to get some feedback. I would prefer not to expose types from other libraries, especially if they are not 1.0 yet. It would be unfortunate if another crates type broke  our compatibility.

Reply:
Ok.  I'll accept this code if you open a bug saying that upload() must be fixed to not load the entire payload into memory.

Reply:
Ack, opened #6